batcher/aligned-batcher/src/lib.rs

```diff
@@ -453,146 +452,168 @@ impl Batcher {
         }
 
         info!("Verifying message signature...");
-        if let Ok(addr) = client_msg.verify_signature() {
-            info!("Message signature verified");
-            if self.is_nonpaying(&addr) {
-                self.handle_nonpaying_msg(ws_conn_sink.clone(), client_msg)
-                    .await
-            } else {
-                if !self
-                    .check_user_balance_and_increment_proof_count(&addr)
-                    .await
-                {
-                    send_message(
-                        ws_conn_sink.clone(),
-                        ValidityResponseMessage::InsufficientBalance(addr),
-                    )
-                    .await;
+        let Ok(addr) = client_msg.verify_signature() else {
+            error!("Signature verification error");
+            send_message(
+                ws_conn_sink.clone(),
+                ValidityResponseMessage::InvalidSignature,
+            )
+            .await;
+            return Ok(());
+        };
+        info!("Message signature verified");
 
-                    return Ok(());
-                }
+        let is_non_paying = self.is_nonpaying(&addr);
+        if is_non_paying {
+            debug!("Non paying message");
+        }
 
-                let nonced_verification_data = client_msg.verification_data;
-                if nonced_verification_data.verification_data.proof.len() > self.max_proof_size {
-                    error!("Proof size exceeds the maximum allowed size.");
-                    send_message(ws_conn_sink.clone(), ValidityResponseMessage::ProofTooLarge)
-                        .await;
-                    return Ok(());
-                }
+        debug!("Verifying balance");
+        if is_non_paying {
+            let non_paying_config = self.non_paying_config.as_ref().unwrap();
+            let addr = non_paying_config.replacement.address();
+            let user_balance = self.get_user_balance(&addr).await;
 
-                // When pre-verification is enabled, batcher will verify proofs for faster feedback with clients
-                if self.pre_verification_is_enabled
-                    && !zk_utils::verify(&nonced_verification_data.verification_data).await
-                {
-                    error!("Invalid proof detected. Verification failed.");
-                    send_message(ws_conn_sink.clone(), ValidityResponseMessage::InvalidProof).await;
-                    return Ok(()); // Send error message to the client and return
-                }
+            if user_balance == U256::from(0) {
```

Comment:
I don't understand why here we check if the balance is zero instead of checking for the actual cost of the verification.